### PR TITLE
feat: support "se:cdp" capability

### LINKED ISSUE #1063: Add Selenium-like websocket endpoint for CDP proxy
Starting from 4.06a Selenium supports CDP proxy and allows connecting websocket to `ws://hub:port/session/${sessionId}/se/cdp`. 
It would be great to have the same interface for Selenoid in addition/instead of existed `ws://selenoid.example.com:4444/devtools/<session-id>/<method>`. The unified interface simplifies writing client libraries.

### PR DESCRIPTION
**Pull request by DudaGod**
Implement `se:cdp` capability to be able to connect from playwright. Related PR to pwt - https://github.com/microsoft/playwright/pull/23341.

fixes #1063

### CODE REVIEW

**Comment by DudaGod** [comment]
Fixed

File: selenoid.go [hunk 1]
```diff
@@ -393,6 +409,28 @@ func removeSelenoidOptions(input []byte) []byte {
...
+			if raw, ok := v["capabilities"]; ok {
+				if c, ok := raw.(map[string]interface{}); ok {
+					c["se:cdp"] = fmt.Sprintf("ws://%s/devtools/%s/", host, v["sessionId"])
+				}
+			}
+		}
+	}
+	ret, err := json.Marshal(body)
+	if err != nil {
+		return ret, fmt.Errorf("stringify body response: %v", err)
```
Comment (vania-pooh) [inline selenoid.go:429]: return nil, fmt.Errorf()

File: selenoid.go [hunk 2]
```diff
@@ -393,6 +409,28 @@ func removeSelenoidOptions(input []byte) []byte {
 	return ret
 }
 
+func addSeCdpCapability(input []byte, host string) ([]byte, error) {
+	body := make(map[string]interface{})
+	err := json.Unmarshal(input, &body)
+	if err != nil {
+		return input, fmt.Errorf("parse body response: %v", err)
```
Comment (vania-pooh) [inline selenoid.go:416]: return nil, fmt.Errorf()

File: selenoid.go [hunk 3]
```diff
@@ -281,6 +281,22 @@ func create(w http.ResponseWriter, r *http.Request) {
...
+			return
+		}
+		newBody, err := addSeCdpCapability(body, r.Host)
+		if err != nil {
+			log.Printf("[%d] [ERROR_ADDING_SE_CDP_CAPABILIITY] [%v]", requestId, err)
+			queue.Drop()
+			cancel()
+			return
+		}
+		resp.Body = io.NopCloser(bytes.NewBufferString(string(newBody)))
```
Comment (vania-pooh) [inline selenoid.go:298]: change to bytes.NewReader(newBody)

**Review by vania-pooh — CHANGES_REQUESTED** [review]

File: selenoid.go [hunk 4]
```diff
@@ -281,6 +281,22 @@ func create(w http.ResponseWriter, r *http.Request) {
...
+		newBody, err := addSeCdpCapability(body, r.Host)
+		if err != nil {
+			log.Printf("[%d] [ERROR_ADDING_SE_CDP_CAPABILIITY] [%v]", requestId, err)
+			queue.Drop()
+			cancel()
+			return
+		}
+		resp.Body = io.NopCloser(bytes.NewReader(newBody))
+		resp.ContentLength = int64(len(newBody))
 		tee := io.TeeReader(resp.Body, w)
```
Comment (vania-pooh) [inline selenoid.go:300]: Probably need to use `body` or `newBody` instead of `resp.Body` here to prevent using already read io.Reader.

**Review by vania-pooh — CHANGES_REQUESTED** [review]

File: selenoid.go (hunk 4 above)
Comment (vania-pooh) [reply selenoid.go:300]: Or you could simply get ID field from already parsed body.

**Review by vania-pooh — APPROVED** [review]

**Comment by martingrossmann** [comment]
@vania-pooh Hi, is'nt this change worth a new release?

**Comment by vania-pooh** [comment]
@martingrossmann will release during next days.

**Pull request merged**